### PR TITLE
Allow a task to cancel itself and add new tasks safely

### PR DESCRIPTION
**Pull request by philj404**
This change permits tasks to cancel themselves without side effects.

### CODE REVIEW

File: extras/tests/unitTest/unitTest.ino [hunk 1]
```diff
@@ -310,6 +310,65 @@ test(timer_cancel_all) {
...
+bool tickB(void *) {
+    ticksB++;
+    if (myTimer) {
+        myTimer->cancel(pendingTask);
+        pendingTask = myTimer->in(1, tickA);  // B then A
+    }
+    return false;
+}
+
+test(timer_cancel_in_task) {
```
Comment (philj404) [inline extras/tests/unitTest/unitTest.ino:340]: This regression test will only PASS with the change (legacy code will fail).  It shows `timer.cancel(myRunningTask)` can work and added tasks will run too.

File: src/arduino-timer.h [hunk 2]
```diff
@@ -131,10 +131,14 @@ class Timer {
                 const unsigned long duration = t - task->start;
 
                 if (duration >= task->expires) {
-                    task->repeat = task->handler(task->opaque) && task->repeat;
-
-                    if (task->repeat) task->start = t;
-                    else remove(task);
+                    Task activeId = task_id(task);
```
Comment (philj404) [inline src/arduino-timer.h:134]: By caching the task's `task_id(task)` and `task->repeat` states, it becomes safe to ignore whether the handler deletes the task while it is running.

File: src/arduino-timer.h [hunk 3]
```diff
@@ -131,10 +131,14 @@ class Timer {
...
+                    bool repeating = task->repeat;
+                    repeating &= task->handler(task->opaque);
+
+                    if (activeId == task_id(task)) {
```
Comment (philj404) [inline src/arduino-timer.h:138]: If this task was deleted (early), the timer should not delete a new task using the slot.

**Comment by contrem** [comment]
My objection to this strategy is the use of `task_id` in the core tick loop. `task_id` is a wrapable value that can collide, and using it in the tick loop if a task is self-canceled violates one of the collision assumptions noted in 114e4046dcfac557daec620d9b3697a34617b70b:

`An active timer with the same count will still occupy its slot.`

It is then possible for a collision to occur during tick processing, creating the same situation this pr tries to avoid, but much harder to diagnose.

**Comment by philj404** [comment]
Good point.  Making the code work "almost all the time" is different from working "every time".  The causes of those problems are very hard to track down.

I have an idea.  I will see if I can create a "show and tell" for a workaround that does not need to modify the library.

**Pull request closed without merging**